From 70ddb5e5d38d3d4215cbf492965d09e062c53c46 Mon Sep 17 00:00:00 2001
From: Matthieu Huin <mhuin@redhat.com>
Date: Wed, 3 Jul 2019 15:01:44 +0000
Subject: [PATCH 1/1] Fix some bazel related problems:

* NoClassDefFoundError: javax/annotation/Generated -> see https://github.com/google/auto/issues/560 and specifically this patch http://paste.openstack.org/show/628224/
* Fix Math round truncation error flagged by error-prone -> backport of https://phabricator.wikimedia.org/rGERRITDEPLOY2a107d74a134cbf757584606f770dc26f82ff1f6
* limit build jobs to 1 to avoid out-of-memory errors
---
 WORKSPACE                                          | 11 +++++++++--
 .../java/com/google/gerrit/client/FormatUtil.java  | 22 +++++++++++++++++++++-
 lib/auto/BUILD                                     | 13 ++++++++++---
 tools/bazel.rc                                     |  2 +-
 4 files changed, 41 insertions(+), 7 deletions(-)

diff --git a/WORKSPACE b/WORKSPACE
index 7641e80..3991cb9 100644
--- a/WORKSPACE
+++ b/WORKSPACE
@@ -434,11 +434,18 @@ maven_jar(
 
 maven_jar(
     name = "auto_value",
-    artifact = "com.google.auto.value:auto-value:1.4",
-    sha1 = "6d1448fcd13074bd3658ef915022410b7c48343b",
+    artifact = "com.google.auto.value:auto-value:1.5",
+    sha1 = "ed31b6bc2e3723c26ea86439862d12ad311b64b3",
 )
 
 maven_jar(
+    name = "javax_annotation",
+    artifact = "javax.annotation:javax.annotation-api:1.3.1",
+    sha1 = "20a2c0583598d68b0835474bbe07792d4f3b219f",
+)
+
+
+maven_jar(
     name = "tukaani_xz",
     artifact = "org.tukaani:xz:1.4",
     sha1 = "18a9a2ce6abf32ea1b5fd31dae5210ad93f4e5e3",
diff --git a/gerrit-gwtui/src/main/java/com/google/gerrit/client/FormatUtil.java b/gerrit-gwtui/src/main/java/com/google/gerrit/client/FormatUtil.java
index 7307264..81672a3 100644
--- a/gerrit-gwtui/src/main/java/com/google/gerrit/client/FormatUtil.java
+++ b/gerrit-gwtui/src/main/java/com/google/gerrit/client/FormatUtil.java
@@ -137,7 +137,27 @@ public class FormatUtil {
     if (size == 0) {
       return Resources.C.notAvailable();
     }
-    int p = Math.abs(Math.round(delta * 100 / size));
+    int p = Math.abs(saturatedCast(delta * 100 / size));
+    // int p = Math.abs(Math.round(delta * 100 / size));
     return p + "%";
   }
+
+  /**
+   * Returns the {@code int} nearest in value to {@code value}.
+   *
+   * @param value any {@code long} value
+   * @return the same value cast to {@code int} if it is in the range of the {@code int} type,
+   *     {@link Integer#MAX_VALUE} if it is too large, or {@link Integer#MIN_VALUE} if it is too
+   *     small
+   */
+  private static int saturatedCast(long value) {
+    if (value > Integer.MAX_VALUE) {
+      return Integer.MAX_VALUE;
+    }
+    if (value < Integer.MIN_VALUE) {
+      return Integer.MIN_VALUE;
+    }
+    return (int) value;
+  }
+
 }
diff --git a/lib/auto/BUILD b/lib/auto/BUILD
index 569398e..52597d6 100644
--- a/lib/auto/BUILD
+++ b/lib/auto/BUILD
@@ -1,13 +1,20 @@
 java_plugin(
     name = "auto-annotation-plugin",
     processor_class = "com.google.auto.value.processor.AutoAnnotationProcessor",
-    deps = ["@auto_value//jar"],
+    deps = ["@auto_value//jar",
+            ":javax-annotation"],
 )
 
 java_plugin(
     name = "auto-value-plugin",
     processor_class = "com.google.auto.value.processor.AutoValueProcessor",
-    deps = ["@auto_value//jar"],
+    deps = ["@auto_value//jar", ":javax-annotation"],
+)
+
+java_library(
+    name = "javax-annotation",
+    data = ["//lib:LICENSE-Apache2.0"],
+    exports = ["@javax_annotation//jar"],
 )
 
 java_library(
@@ -18,5 +25,5 @@ java_library(
         ":auto-value-plugin",
     ],
     visibility = ["//visibility:public"],
-    exports = ["@auto_value//jar"],
+    exports = ["@auto_value//jar", ":javax-annotation",],
 )
diff --git a/tools/bazel.rc b/tools/bazel.rc
index ab974d9..39552cd 100644
--- a/tools/bazel.rc
+++ b/tools/bazel.rc
@@ -1,2 +1,2 @@
-build --workspace_status_command=./tools/workspace-status.sh --strategy=Closure=worker
+build --jobs=1 --workspace_status_command=./tools/workspace-status.sh --strategy=Closure=worker
 test --build_tests_only
-- 
1.8.3.1

